src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Order/TreeItemOrderPropertyProvider.cs

```diff
@@ -0,0 +1,86 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.IO;
+using System.Linq;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS.Tree.Order
+{
+    /// <summary>
+    /// Provider that computes display order of tree items based on input ordering of
+    /// evaluated includes from the project file.
+    /// </summary>
+    internal class TreeItemOrderPropertyProvider : IProjectTreePropertiesProvider
+    {
+        private readonly char[] _separators = new char[]
+        {
+            Path.DirectorySeparatorChar, Path.AltDirectorySeparatorChar
+        };
+
+        public IReadOnlyCollection<ProjectItemIdentity> OrderedItems { get; }
+
+        public ImmutableHashSet<string> AllowedItemTypes { get; }
+
+        public Dictionary<string, int> DisplayOrderMap { get; } = new Dictionary<string, int>(StringComparer.OrdinalIgnoreCase);
+
+        public TreeItemOrderPropertyProvider(IReadOnlyCollection<ProjectItemIdentity> orderedItems)
+        {
+            OrderedItems = orderedItems;
+
+            AllowedItemTypes = OrderedItems.Select(p => p.ItemType).ToImmutableHashSet();
+
+            ComputeIndices(orderedItems);
+        }
+
+        /// <summary>
+        /// Preorder folders and items that are provided as ordered evaluated includes
+        /// </summary>
+        /// <param name="orderedItems">ordered evaluated includes</param>
+        private void ComputeIndices(IReadOnlyCollection<ProjectItemIdentity> orderedItems)
+        {
+            var index = 1;
+            var parts = OrderedItems.SelectMany(p => p.EvaluatedInclude.Split(_separators));
```

Comment:
I'm confused by this - if you have something like "A\B.cs", "C\D\E.cs", aren't you adding "A", "B.cs", "C", "D", and "E.cs" all at the same level?  What if there is an E.cs in multiple folders?

Reply:
+1

---
In reply to: [148030914](https://github.com/dotnet/project-system/pull/2930#discussion_r148030914) [](ancestors = 148030914)

Reply:
With "A\B.cs", "C\D\E.cs", display order map contains { "A":1, "B.cs":2, "C":3, "D":4, and "E.cs":5 } following this method. As I note above, CalculatePropertyValues does not provide the full evaluated include. We can improve this implementation by modifying the data flow to return the original include with each item, so we can distinguish between duplicates.

---
In reply to: [148051124](https://github.com/dotnet/project-system/pull/2930#discussion_r148051124) [](ancestors = 148051124,148030914)

Reply:
So, what is the user experience when there are duplicate short names in different folders?  The one in the second folder ends up sorted before everything else in its folder, even if that's not the order it appears in the .csproj?

Reply:
@Pilchie I've pushed a new commit (https://github.com/dotnet/project-system/pull/2930/commits/3f54d56d6b041d5c4386ec19cb0039881c9f3931) with changes to handle duplicate file names by storing the rooted path for those files and using it to disambiguate. With this fix, local.fs is properly ordered in the following example, instead of appearing first the second time:

![image](https://user-images.githubusercontent.com/7732033/32354380-b4527bf2-bfe6-11e7-8120-d3ca541e09da.png)

It's a bit more involved but I tried to restrict the new code path to only be hit when there are duplicates.